### PR TITLE
feat: use polymorphic scale client

### PR DESCRIPTION
**Pull request by mcavoyk**
PR for https://github.com/kubernetes-sigs/cluster-proportional-autoscaler/issues/91.

Leaves backwards compatibility with targets `replicaset/*` , `deployment/*`, etc, without a given apiGroup, but target can now be any resource `resource.group/*` which implements the scale sub-resource.

### CODE REVIEW

**Comment by MrHohn** [comment]
@mcavoyk Thanks for PR and sorry for my late reply. To make the actual changes more clear, is it possible for you to split this into two commits? One for the CPA code changes and one for the vendor files update? Thank you.

**Comment by mcavoyk** [comment]
Sure, vendor is in its own commit now.

File: pkg/autoscaler/k8sclient/k8sclient.go [hunk 1]
```diff
@@ -81,8 +87,28 @@ func NewK8sClient(namespace, target string, nodelabels string) (K8sClient, error
...
+		return nil, err
+	}
+	glog.Infof("Running with Kubernetes cluster version: v%s.%s. git version: %s. git tree state: %s. commit: %s. platform: %s",
+		v.Major, v.Minor, v.GitVersion, v.GitTreeState, v.GitCommit, v.Platform)
+	glog.Infof("Communication with server successful")
+
+	resolver := scale.NewDiscoveryScaleKindResolver(discoveryClient)
+	cachedDiscoveryClient := cacheddiscovery.NewMemCacheClient(discoveryClient)
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedDiscoveryClient)
+	scaleNamespacer := scale.New(restClient, mapper, dynamic.LegacyAPIPathResolverFunc, resolver)
```
Comment (MrHohn) [inline pkg/autoscaler/k8sclient/k8sclient.go:109]: I'm not very familiar with this API - do you foresee any problem similar to https://github.com/kubernetes-sigs/cluster-proportional-vertical-autoscaler/issues/35? I do see the implementations are different, just not sure about how the underlying discovery will happen.

**Review by MrHohn — APPROVED** [review]
Thank you and sorry for the huge delay on this PR, this looks great - I like that it is able to get rid of a whole lot of codes that handle replica number update. Added a question about API discovery.

**Pull request closed without merging**